op-chain-ops/script/forking/rpc.go

```diff
@@ -0,0 +1,140 @@
+package forking
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/holiman/uint256"
+
+	"github.com/ethereum/go-ethereum"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/common/hexutil"
+
+	"github.com/ethereum-optimism/optimism/op-service/retry"
+)
+
+type RPCClient interface {
+	CallContext(ctx context.Context, result any, method string, args ...any) error
+}
+
+type RPCSource struct {
+	stateRoot common.Hash
+	blockHash common.Hash
+
+	maxAttempts int
+	timeout     time.Duration
+	strategy    retry.Strategy
+
+	ctx    context.Context
+	cancel context.CancelFunc
+
+	client     RPCClient
+	urlOrAlias string
+}
+
+var _ ForkSource = (*RPCSource)(nil)
```

Comment:
Foundry does a thing where it caches locally on the filesystem for remote calls so its only slow the first time that it runs. Not sure if we need to go that far in the first pass, could be good to have a cache client that does that all in memory, that would get things like `block.number`

Reply:
proto added this